contracts/utils/cryptography/MerkleProof.sol

```diff
@@ -23,14 +23,63 @@ library MerkleProof {
         bytes32 root,
         bytes32 leaf
     ) internal pure returns (bool) {
+        (bool success, ) = verifyAndRecoverIndex(proof, root, leaf);
+        return success;
+    }
+
+    /**
+     * @dev Returns true if a `leaf` can be proved to be a part of a Merkle tree
+     * defined by `root`. For this, a `proof` must be provided, containing
+     * sibling hashes on the branch from the leaf to the root of the tree. Each
+     * pair of leaves and each pair of pre-images are assumed to be sorted.
+     */
+    function verifyAndRecoverIndex(
+        bytes32[] memory proof,
+        bytes32 root,
+        bytes32 leaf
+    ) internal pure returns (bool, uint256) {
         bytes32 computedHash = leaf;
+        uint256 index = 0;
 
         for (uint256 i = 0; i < proof.length; i++) {
             bytes32 proofElement = proof[i];
 
             if (computedHash <= proofElement) {
                 // Hash(current computed hash + current element of the proof)
                 computedHash = keccak256(abi.encodePacked(computedHash, proofElement));
+            } else {
+                // Hash(current element of the proof + current computed hash)
+                computedHash = keccak256(abi.encodePacked(proofElement, computedHash));
+                index |= 1 << i;
```

Comment:
Note: this indexing me logic doesn't guarantee uniqueness of indexes because proofs are processed from the leaf up to the root, and not all branches have equal length.

Example:
```
   Root
   /  \
  X    C
 / \
A   B
```

Path from B to root is :
- right for B to X →  `index |= 1 << 0`
- left from X to root  → no change
in the end, `index = 1`

Path from C to root:
- right from C to root  →  `index |= 1 << 0`
in the end, `index = 1`

This can be fixed by using `index |= 1 << (proof.length -1)` or using the solution proposed in #2841

Reply:
AFAIK, there is no simple way to retrieve the leaf index with the information we have here.

Reply:
@Amxx I see, will think of it.

Reply:
The only solution I see here is to add one more argument: `treeHeight`

Reply:
It could and should be stored in `immutable` paired with `merkleRoot`